internal/actionwait/wait.go

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+// Package actionwait provides a lightweight, action-focused polling helper
+// for imperative Terraform actions which need to await asynchronous AWS
+// operation completion with periodic user progress events.
+package actionwait
+
+import (
+	"context"
+	"errors"
+	"slices"
+	"strings"
+	"time"
+)
+
+// DefaultPollInterval is the default fixed polling interval used when no custom IntervalStrategy is provided.
+const DefaultPollInterval = 30 * time.Second
+
+// Status represents a string status value returned from a polled API.
+type Status string
+
+// FetchResult wraps the latest status (and optional value) from a poll attempt.
+// Value may be a richer SDK structure (pointer) or zero for simple cases.
+type FetchResult[T any] struct {
+	Status Status
+	Value  T
+}
+
+// FetchFunc retrieves the latest state of an asynchronous operation.
+// It should be side-effect free aside from the remote read.
+type FetchFunc[T any] func(context.Context) (FetchResult[T], error)
+
+// IntervalStrategy allows pluggable poll interval behavior (fixed, backoff, etc.).
+type IntervalStrategy interface { //nolint:interfacebloat // single method interface (tiny intentional interface)
+	NextPoll(attempt uint) time.Duration
+}
+
+// FixedInterval implements IntervalStrategy with a constant delay.
+type FixedInterval time.Duration
+
+// NextPoll returns the fixed duration.
+func (fi FixedInterval) NextPoll(uint) time.Duration { return time.Duration(fi) }
+
+// Options configure the WaitForStatus loop.
+type Options[T any] struct {
+	Timeout            time.Duration    // Required total timeout.
+	Interval           IntervalStrategy // Poll interval strategy (default: 30s fixed).
+	ProgressInterval   time.Duration    // Throttle for ProgressSink (default: disabled if <=0).
+	SuccessStates      []Status         // Required (>=1) terminal success states.
+	TransitionalStates []Status         // Optional allowed in-flight states.
+	FailureStates      []Status         // Optional explicit failure states.
+	ConsecutiveSuccess int              // Number of consecutive successes required (default 1).
+	ProgressSink       func(fr FetchResult[any], meta ProgressMeta)
+}
+
+// ProgressMeta supplies metadata for progress callbacks.
+type ProgressMeta struct {
+	Attempt    uint
+	Elapsed    time.Duration
+	Remaining  time.Duration
+	Deadline   time.Time
+	NextPollIn time.Duration
+}
+
+// TimeoutError is returned when the operation does not reach a success state within Timeout.
+type TimeoutError struct {
```

Comment:
It may be "nice" to consolidate these errors with those in `internal/retry` and maybe move them all to a more generic package, e.g. `internal/errs`?